### PR TITLE
Added Chatwheel Sync Final

### PR DESCRIPTION
**Pull request by numlck**
Still have to update the 12v12 Mode if this is accepted.

Working and tested in game,

Debounced + Batched Updates in realtime.

### CODE REVIEW

File: Server/Controllers/MatchController.cs [hunk 1]
```diff
@@ -271,8 +291,30 @@ public async Task<List<object>> Events([FromBody] MatchEventsRequest request)
 
             return events.Select(e => e.Body).ToList();
         }
+
+
     }
+        public class PatreonPlayerUpdate {
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs]:
```suggestion

    public class PatreonPlayerUpdate {
```

File: Server/Controllers/MatchController.cs [hunk 2]
```diff
@@ -271,8 +291,30 @@ public async Task<List<object>> Events([FromBody] MatchEventsRequest request)
...
+        [Required] public CustomGame? CustomGame { get; set; }
+        [Required] public long MatchId { get; set; }
+        [Required] public string MapName { get; set; }
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs]: All of these are unused

File: Server/Controllers/MatchController.cs [hunk 3]
```diff
@@ -271,8 +291,30 @@ public async Task<List<object>> Events([FromBody] MatchEventsRequest request)
...
+        [Required] public List<Player> players { get; set; }
+        public class Player
+        {
+            [Required] public ushort PlayerId { get; set; }
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs]: That one is unused too

File: Server/appsettings.Development.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
     }
   },
   "ConnectionStrings": {
-    "Database": "Server=127.0.0.1;Port=5432;Database=postgres;User Id=postgres;Password=postgres"
+    "Database": "Server=127.0.0.1;Port=5432;Database=postgres;User Id=postgres;Password=0000"
```
Comment (ark120202) [inline Server/appsettings.Development.json]: These are the parameters used by the docker container, you shouldn't change them

File: Server/Controllers/MatchController.cs [hunk 5]
```diff
@@ -258,7 +258,27 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
...
+                Console.WriteLine("Updating Player " + player.SteamId);
+                ctx.PatreonBootsEnabled = player.PatreonUpdate.BootsEnabled;
+                ctx.PatreonEmblemEnabled = player.PatreonUpdate.EmblemEnabled;
+                ctx.PatreonEmblemColor = player.PatreonUpdate.EmblemColor;
+                ctx.PatreonChatWheelFavorites = player.PatreonUpdate.ChatWheelFavorites;
+                
+            }
+            await _context.SaveChangesAsync();
+            return Ok();
+        }
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs:283]:
```suggestion
        }

```

File: Server/Controllers/MatchController.cs [hunk 6]
```diff
@@ -258,7 +258,27 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
...
+
+                _context.Players.Update(ctx);
+                
+                Console.WriteLine("Updating Player " + player.SteamId);
+                ctx.PatreonBootsEnabled = player.PatreonUpdate.BootsEnabled;
+                ctx.PatreonEmblemEnabled = player.PatreonUpdate.EmblemEnabled;
+                ctx.PatreonEmblemColor = player.PatreonUpdate.EmblemColor;
+                ctx.PatreonChatWheelFavorites = player.PatreonUpdate.ChatWheelFavorites;
+                
+            }
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs:280]:
```suggestion
            }

```

File: Server/Controllers/MatchController.cs [hunk 7]
```diff
@@ -258,7 +258,27 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
...
+        public async Task<ActionResult> PatreonPlayerUpdate([FromBody] PatreonPlayerUpdate request)
+        {
+            foreach (var player in request.players)
+            {
+                var ctx = _context.Players.FirstOrDefault(p => p.SteamId.ToString() == player.SteamId);
+                if (ctx == null) continue;
+
+                _context.Players.Update(ctx);
+                
+                Console.WriteLine("Updating Player " + player.SteamId);
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs]: Should be removed before merge

File: Server/Controllers/MatchController.cs [hunk 8]
```diff
@@ -258,7 +258,27 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
 
             return Ok();
         }
-
+        [HttpPost]
+        [Route("patreon_player_update")]
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs]: Use kebab-case for routes
```suggestion
        [Route("patreon-player-update")]
```

File: Server/Controllers/MatchController.cs [hunk 9]
```diff
@@ -258,7 +258,27 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
...
+        public async Task<ActionResult> PatreonPlayerUpdate([FromBody] PatreonPlayerUpdate request)
+        {
+            foreach (var player in request.players)
+            {
+                var ctx = _context.Players.FirstOrDefault(p => p.SteamId.ToString() == player.SteamId);
+                if (ctx == null) continue;
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs]: It should create entities for players that don't exist yet, like in the match end event

File: Server/Controllers/MatchController.cs [hunk 10]
```diff
@@ -258,7 +258,27 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
...
+
+                _context.Players.Update(ctx);
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs]: IIRC all entities coming form database operations are already tracked

File: .gitignore [hunk 11]
```diff
@@ -1,8 +1,6 @@
 obj/
 bin/
-
```
Comment (ark120202) [inline .gitignore:3]: Unrelated change

**Comment by numlck** [comment]
PR Fix

File: Server/Controllers/MatchController.cs [hunk 12]
```diff
@@ -258,7 +258,28 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
 
             return Ok();
         }
-
+        [HttpPost]
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs:262]:
```suggestion

        [HttpPost]
```

File: Server/Controllers/MatchController.cs [hunk 13]
```diff
@@ -258,7 +258,28 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
...
+        [Route("patreon-player-update")]
+        public async Task<ActionResult> PatreonPlayerUpdate([FromBody] PatreonPlayerUpdate request)
+        {
+            foreach (var player in request.players)
+            {
+                var ctx = _context.Players.FirstOrDefault(p => p.SteamId.ToString() == player.SteamId);
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs]: `ctx` isn't really what it is, plus that's an abbreviation that makes code harder to read. It's better 
 to rename it to `player` and name the loop variable `playerUpdate`

File: Server/Controllers/MatchController.cs [hunk 14]
```diff
@@ -272,7 +293,24 @@ public async Task<List<object>> Events([FromBody] MatchEventsRequest request)
...
         }
     }
+    public class PatreonPlayerUpdate {
+
+        [Required] public List<Player> players { get; set; }
+        public class Player
+        {
+            [Required] public string SteamId { get; set; }
 
+            public PatreonUpdate PatreonUpdate { get; set; }
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs]: I guess it should be required?

File: Server/Controllers/MatchController.cs [hunk 15]
```diff
@@ -272,7 +293,24 @@ public async Task<List<object>> Events([FromBody] MatchEventsRequest request)
             return events.Select(e => e.Body).ToList();
         }
     }
+    public class PatreonPlayerUpdate {
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs:299]:
```suggestion

    public class PatreonPlayerUpdate {
```

File: Server/Controllers/MatchController.cs [hunk 16]
```diff
@@ -272,7 +293,24 @@ public async Task<List<object>> Events([FromBody] MatchEventsRequest request)
...
+        }
+        public class PatreonUpdate
+        {
+            public bool EmblemEnabled { get; set; }
+            public string EmblemColor { get; set; }
+            public bool BootsEnabled { get; set; }
+            // TODO: Required?
+            public List<int>? ChatWheelFavorites { get; set; }
+        }
+    }
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs:315]:
```suggestion
    }

```

File: Server/Controllers/MatchController.cs [hunk 17]
```diff
@@ -272,7 +293,24 @@ public async Task<List<object>> Events([FromBody] MatchEventsRequest request)
...
+    public class PatreonPlayerUpdate {
+
+        [Required] public List<Player> players { get; set; }
+        public class Player
+        {
+            [Required] public string SteamId { get; set; }
 
+            public PatreonUpdate PatreonUpdate { get; set; }
+        }
+        public class PatreonUpdate
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs:308]:
```suggestion
        }

        public class PatreonUpdate
```

File: Server/appsettings.Development.json [hunk 18]
```diff
@@ -7,7 +7,7 @@
     }
   },
   "ConnectionStrings": {
-    "Database": "Server=127.0.0.1;Port=5432;Database=postgres;User Id=postgres;Password=postgres"
+    "Database": "Server=database;Port=5432;Database=postgres;User Id=postgres;Password=postgres"
```
Comment (ark120202) [inline Server/appsettings.Development.json:10]: It should not be changed

File: Server/appsettings.json [hunk 19]
```diff
@@ -6,6 +6,6 @@
   },
   "AllowedHosts": "*",
   "ConnectionStrings": {
-    "Database": "Server=database;Port=5432;Database=postgres;User Id=postgres;Password=postgres"
+    "Database": "Server=database;Port=5432;Database=postgres;UserId=postgres;Password=postgres"
```
Comment (ark120202) [inline Server/appsettings.json:9]: Same here

**Review by ark120202 — COMMENTED** [review]
Also see few unresolved comments above

File: Server/Controllers/MatchController.cs [hunk 20]
```diff
@@ -273,6 +296,24 @@ public async Task<List<object>> Events([FromBody] MatchEventsRequest request)
...
     }
 
+    public class PatreonPlayerUpdate {
+
+        [Required] public List<Player> players { get; set; }
+        public class Player
+        {
+            [Required] public string SteamId { get; set; }
+
+            [Required] public PatreonUpdate PatreonUpdate { get; set; }
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs:306]:
```suggestion
            [Required] public PatreonUpdate PatreonUpdate { get; set; }
```

File: Server/Controllers/MatchController.cs [hunk 21]
```diff
@@ -259,6 +259,29 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
             return Ok();
         }
 
+        [HttpPost]
+        [Route("patreon-player-update")]
+        public async Task<ActionResult> PatreonPlayerUpdate([FromBody] PatreonPlayerUpdate request)
+        {
+            foreach (var player_update in request.players)
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs:266]: Local identifiers should be in camelCase
```suggestion
            foreach (var playerUpdate in request.players)
```

File: Server/Controllers/MatchController.cs [hunk 22]
```diff
@@ -259,6 +259,29 @@ public async Task<ActionResult> After([FromBody] AfterMatchRequest request)
...
+            {
+                var player = _context.Players.FirstOrDefault(p => p.SteamId.ToString() == player_update.SteamId);
+                if (player == null)
+                {
+                    player = new Player() { SteamId = ulong.Parse(player_update.SteamId) };
```
Comment (ark120202) [inline Server/Controllers/MatchController.cs:271]:
```suggestion
				var steamId = ulong.Parse(playerUpdate.SteamId);
                var player = _context.Players.FirstOrDefault(p => p.SteamId == steamId );
                if (player == null)
                {
                    player = new Player() { SteamId = steamId };
```
